### PR TITLE
Using node name to improve node controller performance. 

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

Part of #67823

```release-note
None
```

### CODE REVIEW

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -353,8 +354,10 @@ func NewNodeLifecycleController(podInformer coreinformers.PodInformer,
 				nc.nodeUpdateQueue.Add(node)
 				return nil
 			}),
-			UpdateFunc: nodeutil.CreateUpdateNodeHandler(func(_, newNode *v1.Node) error {
-				nc.nodeUpdateQueue.Add(newNode)
+			UpdateFunc: nodeutil.CreateUpdateNodeHandler(func(oldNode, newNode *v1.Node) error {
+				if !reflect.DeepEqual(oldNode.Status.Conditions, newNode.Status.Conditions) {
```
Comment (shyamjvs) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: I'm not sure if this is correct, as IIUC this means we would never change the taints if we failed to do it initially and the node conditions didn't change ever after that.

@wojtek-t

Comment (wojtek-t) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: So you're saying - if we fail all attempts of retrying, then we won't restart.

Yes - we should readd that node to the queue in that case.

Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: correct !

Comment (shyamjvs) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: > So you're saying - if we fail all attempts of retrying, then we won't restart.

Yes

**Comment by k82cn** [comment]
> I'm not sure I understand how this improves any behaviour - could you clarify?

`FIFO` is implemented by map, so there at most one Node (the latest one from update) in the queue to handle. `workqueue` is implemented by slice which will be several pending node here.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 2]
```diff
@@ -223,7 +222,7 @@ type Controller struct {
 	taintNodeByCondition bool
 
 	nodeUpdateChannels []chan *v1.Node
-	nodeUpdateQueue    workqueue.Interface
+	nodeUpdateQueue    *cache.FIFO
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: I'm fine with trying that, but please add a comment so that we know what this was changed.

**Comment by shyamjvs** [comment]
We can try, but I'm still not 100% sure if this will help much. As IIUC this is only making the goroutine that dispatches the items to channels (which I feel is quite fast) to avoid having more than one instance of a node at a time. We may still end up with as many workitems in the hashed channel as before (which is where I think more time is consumed). Thoughts?

**Comment by wojtek-t** [comment]
> We can try, but I'm still not 100% sure if this will help much. As IIUC this is only making the goroutine that dispatches the items to channels (which I feel is quite fast) to avoid having more than one instance of a node at a time. We may still end up with as many workitems in the hashed channel as before (which is where I think more time is consumed). Thoughts?

If we are accumulating backlog currently (which we potentially can?) that will reduce the latency to process a given item.

**Comment by shyamjvs** [comment]
Right, but if we're accumulating backlog in the channels and not in the workqueue which dispatches items to channels? In that case this change may not help much fmu.

**Comment by wojtek-t** [comment]
> Right, but if we're accumulating backlog in the channels and not in the workqueue which dispatches items to channels? In that case this change may not help much fmu.

This has a potential to help (though we need to verify that).
Without this change, you have to process every single change to a node, with that change, you process only the most recent one.

So imagine that your workers have throughput of 100items/s.
if the incoming throughput is 500items/s, that means that you're accumulating backlog over time and will be lagging more and more.

With this change, it's no longer the case, because there is always at most one version of a given object in the queue. So size of backlog is alwasy limitted by 5000 items (for 5000-node cluster), which means, latency for processing a single item (assuming even distribution) is something like 5000items / 100items/s = 50s (instead of indefinitely growing).

**Comment by shyamjvs** [comment]
I do the understand the above reasoning :) Let me clarify what I was saying:

FMU there are currently 2 levels of processing when a node update event comes. First level is the dispatcher - this is a single goroutine which takes the node events and sends it to one of the N channels based on node hash (this is the one where we make sure that only latest one goes). But second level is the workers who process items from their respective channel (and there we can have multiple workitems for the same node - so all of them would be processed anyway). So if the dispatcher is fast and let's say it almost immediately puts the workitems into channels as they come, then all the items go to channels (as it didn't get a chance to buffer for long enough so it could remove older workitems). This can happen if the workers are slower than the dispatcher. So basically,

> With this change, it's no longer the case, because there is always at most one version of a given object in the queue.

In the queue, yes. In the channels, not necessarily.

> So size of backlog is alwasy limitted by 5000 items

Not necessarily (if dispatcher is much faster than workers)

Of course I'm not claiming that the above is indeed the case, we need to verify it.

**Comment by wojtek-t** [comment]
> So if the dispatcher is fast and let's say it almost immediately puts the workitems into channels as they come,

And here is the thing that you miss. The dispatcher is not infinitely fast.
And the reason for that is that the channels (where it puts the work items) have a limited capacity - if they are full (because workers are slow), dispatcher is blocked.

This (it's debatable whether we call it a feature or a bug) is pretty crucial point in my reasoning above.

**Comment by k82cn** [comment]
should we have a try?

**Comment by wojtek-t** [comment]
I'm fine with trying.

/lgtm

**Comment by liggitt** [comment]
/hold

the workqueue will already dedupe... shouldn't we fix this by making the node name the key, rather than enqueing the entire node object?

**Comment by liggitt** [comment]
similar to https://github.com/kubernetes/kubernetes/pull/65350

**Comment by k82cn** [comment]
@liggitt , PTAL :)

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 3]
```diff
@@ -493,15 +493,20 @@ func (nc *Controller) doNoScheduleTaintingPassWorker(i int, stopCh <-chan struct
...
 			}
 		}
 	}
 }
 
-func (nc *Controller) doNoScheduleTaintingPass(node *v1.Node) error {
+func (nc *Controller) doNoScheduleTaintingPass(nodeName string) error {
+	node, err := nc.nodeLister.Get(nodeName)
+	if err != nil {
+		return err
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: if the node is not found by the time we get to this handler, I don't think this needs to return/log an error

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 4]
```diff
@@ -408,16 +408,16 @@ func (nc *Controller) Run(stopCh <-chan struct{}) {
...
+				nodeName := obj.(string)
+				hash := hash(nodeName, scheduler.UpdateWorkerSize)
 
 				select {
 				case <-stopCh:
-					nc.nodeUpdateQueue.Done(node)
+					nc.nodeUpdateQueue.Done(nodeName)
 					return
-				case nc.nodeUpdateChannels[hash] <- node:
+				case nc.nodeUpdateChannels[hash] <- nodeName:
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: What's the reason for making this handling asynchronous? Calling Done prior to actually handling the node in doNoScheduleTaintingPassWorker means the same node can get pushed into the nodeUpdateChannels more than once, right?

Normally, controller just spin up n worker threads to take the next item off a workqueue and process it synchronously. Is there a reason that wouldn't work here, instead of this three level informer->workqueue->channel approach?

**Review by liggitt — COMMENTED** [review]
The change to nodeName as key and lister as lookup in the handler looks good, thanks. 

I did have one question about the additional channel indirect… it seems like that still allows the same node to be queued for processing multiple times, and I'm not seeing the benefit over worker threads handling the workqueue directly.

**Comment by liggitt** [comment]
The IsNotFoundError check aside, this is better than what we have today, so I'd be ok with iterating on it, but I do think the channel indirect worker thread level can/should probably be removed in a follow-up

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 3 above)
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Agree, will update it.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 4 above)
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: There maybe a race condition here: 

1. worker get item from queue
1. worker get item from Lister
1. item was updated during `doNoScheduleTaintingPass`

I just go through workqueue just now: we dedup before `Done`; but there maybe a race condition during `doNoScheduleTaintingPass` node was updated after `Lister.Get`, and we need to always check the latest version of Node. Or we dependent on Node's update next time, and call `Done` after `doNoScheduleTaintingPass`

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 5]
```diff
@@ -493,15 +493,20 @@ func (nc *Controller) doNoScheduleTaintingPassWorker(i int, stopCh <-chan struct
...
+			if err := nc.doNoScheduleTaintingPass(nodeName); err != nil {
+				glog.Errorf("Failed to taint NoSchedule on node <%s>: %v", nodeName, err)
 			}
 		}
 	}
 }
 
-func (nc *Controller) doNoScheduleTaintingPass(node *v1.Node) error {
+func (nc *Controller) doNoScheduleTaintingPass(nodeName string) error {
+	node, err := nc.nodeLister.Get(nodeName)
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: just to double check, doNoScheduleTaintingPass doesn't mutate anything on the node object, right?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 4 above)
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: > Or we dependent on Node's update next time, and call Done after `doNoScheduleTaintingPass`

We can not dependent on that :( , if node is unreachable, we'll not update status.

Comment (liggitt) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: moving all processing of the node to occur before Done() is called would make this simpler to reason about

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 5 above)
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: yes, we only need to get node's condition and taints, and set update request if condition/taint did not match.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 4 above)
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: In this case, I think we still need to use map, e.g. FIFO.

Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: After `Lister.Get`, before `doNoScheduleTaintingPass` finished, there maybe node update events. We can not ignore this event, and it's better to only handle the latest event if multiple.

Comment (liggitt) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: if an update is pushed into the store the lister uses, the informer will be notified, and it will enqueue the node name into the workqueue. the workqueue will only dedupe if the node name is already enqueued and waiting, so we will not ignore the event

Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: oh, got your point; dirty + processing handled that well; thanks .

**Comment by k82cn** [comment]
>  I do think the channel indirect worker thread level can/should probably be removed in a follow-up

Yes, I think it's not necessary; I'll follow up in other PR.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 6]
```diff
@@ -493,15 +491,27 @@ func (nc *Controller) doNoScheduleTaintingPassWorker(i int, stopCh <-chan struct
...
+			if err := nc.doNoScheduleTaintingPass(nodeName); err != nil {
+				glog.Errorf("Failed to taint NoSchedule on node <%s>: %v", nodeName, err)
 			}
 		}
 	}
 }
 
-func (nc *Controller) doNoScheduleTaintingPass(node *v1.Node) error {
+func (nc *Controller) doNoScheduleTaintingPass(nodeName string) error {
+	defer nc.nodeUpdateQueue.Done(nodeName)
```
Comment (k82cn) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: I moved `Done` to here to avoid requeue item during the process. The item (if any update during process) will requeue in `Done`.

Comment (liggitt) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: splitting Get and Done across methods makes this really hard to follow (even this change doesn't properly call done in stopCh cases).

I'd much rather resolve this by getting rid of the async channel handling

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 7]
```diff
@@ -396,36 +395,16 @@ func (nc *Controller) Run(stopCh <-chan struct{}) {
...
 		defer nc.nodeUpdateQueue.ShutDown()
 
 		// Start workers to update NoSchedule taint for nodes.
 		for i := 0; i < scheduler.UpdateWorkerSize; i++ {
-			go nc.doNoScheduleTaintingPassWorker(i, stopCh)
+			// Thanks to "workqueue", each worker just need to get item from queue, because
+			// the item is flagged when got from queue: if new event come, the new item will
+			// be re-queued until "Done", so no more than one worker handle the same item and
+			// no event missed.
+			go nc.doNoScheduleTaintingPassWorker()
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: nit: `go wait.Until(nc.doNoScheduleTaintingPassWorker, time.Second, stopCh)`

**Comment by liggitt** [comment]
nit on using `go wait.Until`, lgtm otherwise. @wojtek-t should probably take another pass as well.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 7 above)
Comment (liggitt) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: wait.Until adds panic handling and will restart the worker until the stop channel is closed in case of panic

Comment (wojtek-t) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: +1

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 8]
```diff
@@ -488,20 +467,35 @@ func (nc *Controller) doFixDeprecatedTaintKeyPass(node *v1.Node) error {
...
-func (nc *Controller) doNoScheduleTaintingPass(node *v1.Node) error {
+func (nc *Controller) doNoScheduleTaintingPass(nodeName string) error {
+
+	node, err := nc.nodeLister.Get(nodeName)
+	if err != nil {
+		// If node not found, just ignore it.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: nit: remove empty line

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 9]
```diff
@@ -488,20 +467,35 @@ func (nc *Controller) doFixDeprecatedTaintKeyPass(node *v1.Node) error {
...
 			return
-		case node := <-nc.nodeUpdateChannels[i]:
-			if err := nc.doNoScheduleTaintingPass(node); err != nil {
-				glog.Errorf("Failed to taint NoSchedule on node <%s>: %v", node.Name, err)
-			}
 		}
+		nodeName := obj.(string)
+
+		if err := nc.doNoScheduleTaintingPass(nodeName); err != nil {
+			glog.Errorf("Failed to taint NoSchedule on node <%s>: %v", nodeName, err)
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: If at some point we decide to not update node status that frequently, this may become a problem (e.g. if you fail to send patch request).

Please add a todo to requeue this node name again in that case.

**Review by wojtek-t — COMMENTED** [review]
Couple minor comments - other than that lgtm.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 7 above)
Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: > keep the doNoScheduleTaintingPassWorker as is with the for loop, 

Oh, yes, there's for loop in `doNoScheduleTaintingPassWorker` :)
Done

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 10]
```diff
@@ -488,20 +467,35 @@ func (nc *Controller) doFixDeprecatedTaintKeyPass(node *v1.Node) error {
...
-		case node := <-nc.nodeUpdateChannels[i]:
-			if err := nc.doNoScheduleTaintingPass(node); err != nil {
-				glog.Errorf("Failed to taint NoSchedule on node <%s>: %v", node.Name, err)
-			}
 		}
+		nodeName := obj.(string)
+
+		if err := nc.doNoScheduleTaintingPass(nodeName); err != nil {
+			// Add nodeName back to the queue, it'll be requeue after `Done`.
+			nc.nodeUpdateQueue.Add(nodeName)
```
Comment (liggitt) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: If you want to requeue, this needs to use the ratelimiting workqueue and track number of retries per item so we don't hotloop on a persistently failing item. 

See use of NewRateLimitingQueue(DefaultControllerRateLimiter()) elsewhere, along with AddRateLimited()/Forget() for handling requeuing items. 

I think the switch to requeue on error can be done as a follow up, since nodes are still updated every few seconds in this release

Comment (k82cn) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: OK, removed .

Comment (wojtek-t) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: +1 to @liggitt - this is non-trivial, that's why i suggested a TODO (and solve it in follow-up).

**Comment by wojtek-t** [comment]
LGTM - thanks!

**Comment by tpepper** [comment]
/priority critical-urgent

This is likely release blocking.
